### PR TITLE
 knot-resolver: add luaPackages.psl 

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Description of changes
This basically adds support for ``policy.slice_randomize_psl()`` 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by vcunat — DISMISSED** [review]
It won't build for me
```
$ nix build -f. luajitPackages.psl

       > Error: Could not find header file for PSL
       >   No file libpsl.h in /nix/store/3gc64ag4cav0y40pk5zd75cs58q25a9l-wrap-lua-hook/include
       >   No file libpsl.h in /nix/store/h523pxxyypw7bk977kirhdzjjv52gzcs-luajit-2.1.1693350652/include
       >   No file libpsl.h in /nix/store/xfc3kaxwvya9i0589b8gn9cnr2h6ah2p-libpsl-0.21.5-bin/include
       > You may have to install PSL in your system and/or pass PSL_DIR or PSL_INCDIR to the luarocks command.
       > Example: luarocks install psl PSL_DIR=/usr/local
```

**Comment by vcunat** [comment]
Note: it's been unclear for us whether that randomized approach improves privacy or makes it worse.  We haven't implemented support for this in the knot-resolver 6.x configuration model so far, BTW.

**Comment by SuperSandro2000** [comment]
The other alternative, to send all requests to all servers, would generate more traffic and we use it more to spread the requests somewhat to all servers.

**Comment by SuperSandro2000** [comment]
I've test this on 23.11 and funnily I think #292260 broke the pkgconfig file.

**Comment by vcunat** [comment]
I'm not aware of such an alternative that increases traffic amount.

**Comment by vcunat** [comment]
Now it looks good to me.  Though the http module won't work until fixes in PR #303527 (guarded by eval error now).

File: pkgs/development/lua-modules/overrides.nix [hunk 1]
```diff
@@ -585,6 +586,15 @@ in
     };
   });
 
+  psl = prev.psl.overrideAttrs (drv: {
+    buildInputs = drv.buildInputs or [ ] ++ [ libpsl ];
+
+    luarocksConfig.variables = {
```
Comment (teto) [inline pkgs/development/lua-modules/overrides.nix:592]: this erases the generated luarocksConfig, do `luarocksConfig = drv.luarocksConfig // { variables = { .. };` or recursive update, look how it's done for other packages.

File: maintainers/scripts/luarocks-packages.csv [hunk 2]
```diff
@@ -110,6 +110,7 @@ nvim-nio,,,,,,mrcjkb
 pathlib.nvim,,,,,,
 penlight,https://github.com/lunarmodules/Penlight.git,,,,,alerque
 plenary.nvim,https://github.com/nvim-lua/plenary.nvim.git,,,,5.1,
+psl,,,,0.3,,
```
Comment (teto) [inline maintainers/scripts/luarocks-packages.csv:113]: do you want to set yourself as maintainer ?

**Comment by SuperSandro2000** [comment]
> I'm not aware of such an alternative that increases traffic amount.

I might have mixed that up with dnsmasq 😅 I did to many DNS things yesterday

File: pkgs/servers/dns/knot-resolver/default.nix [hunk 3]
```diff
@@ -114,7 +114,8 @@ wrapped-full = runCommand unwrapped.name
       # For http module, prefill module, trust anchor bootstrap.
       # It brings lots of deps; some are useful elsewhere (e.g. cqueues).
       http
-      # psl isn't in nixpkgs yet, but policy.slice_randomize_psl() seems not important.
+      # used by policy.slice_randomize_psl()
```
Comment (teto) [inline pkgs/servers/dns/knot-resolver/default.nix:117]: is slice_randomize_psl usage  tested ?I've broken knot-resolver a few times recently so the more we test it, the less worried I am about breaking @vcunat 's favortie software :)

File: pkgs/development/lua-modules/overrides.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/lua-modules/overrides.nix]: 👍🏼 changed

File: maintainers/scripts/luarocks-packages.csv (hunk 2 above)
Comment (SuperSandro2000) [reply maintainers/scripts/luarocks-packages.csv:113]: No, not really. The library seems to be mostly dead, though.

File: pkgs/servers/dns/knot-resolver/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/servers/dns/knot-resolver/default.nix:117]: I think if http is working, this is also working.

Comment (vcunat) [reply pkgs/servers/dns/knot-resolver/default.nix:117]: Sounds safe enough to me, but longterm future/maintenance of PSL stuff seems unclear.

**Comment by vcunat** [comment]
You broke this again.  `luarocksConfig.variables` needs to be used, not `luarocksConfig`.

**Comment by SuperSandro2000** [comment]
> You broke this again. `luarocksConfig.variables` needs to be used, not `luarocksConfig`.

🤦🏼

**Comment by SuperSandro2000** [comment]
Can we merge this?

**Pull request merged**